src/rgw/rgw_gc.cc

```diff
@@ -220,19 +222,31 @@ class RGWGCIOManager {
   }
 
   void schedule_tag_removal(int index, string tag) {
```

Comment:
I would like to see some comments given how convoluted the code has become. So for schedule_tag_removal, I request that you add something along the following lines:
`// This is a request to schedule a tag removal. It will be called once when there are no shadow objects. But it will also be called for every shadow object when there are any. Since we do not want the tag to be removed until all shadow objects have been successfully removed, the scheduling will not happen until the shadow object count goes down to zero.`